### PR TITLE
feat: rewrite dependency resolving and manifest loading

### PR DESCRIPTION
**Pull request by Suyashtnt**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -402,5 +310,118 @@ fn create_manifest_from_config(
...
+                package.package.name
+            )
+        })
+        .change_context(MainError)?;
+
+    let final_package = (loader_data, package);
+    let mut final_vec = vec![final_package];
+    final_vec.append(&mut deps);
+
+    Ok(final_vec)
```
Comment (shaunsingh) [inline src/main.rs:426]: Not stopping the review, but I'd prefer move this logic into its own file and keep the main file pretty small

File: src/manifest.rs [hunk 2]
```diff
@@ -155,3 +187,74 @@ pub struct Plugin {
...
+
+    //     generations_file.add_to_generation(0, PathBuf::from("manifest_path"));
+    //     let latest_generation = generations_file.get_latest_generation_number(0).unwrap();
+    //     assert_eq!(latest_generation, 1);
+
+    //     generations_file.add_to_generation(0, PathBuf::from("manifest_path2"));
+    //     let latest_generation = generations_file.get_latest_generation_number(0).unwrap();
+    //     assert_eq!(latest_generation, 2);
+    // }
+}
```
Comment (shaunsingh) [inline src/manifest.rs:260]: I'd prefer to have a few tests if you can

File: src/smith/git.rs [hunk 3]
```diff
@@ -177,7 +177,7 @@ impl Smith for Git {
             .find_reference("FETCH_HEAD")
             .into_report()
             .change_context(GitError::GitError)
-            .attach_printable_lazy(|| format!("Failed to find FETCH_HEAD: {url}"))
+            .attach_printable_lazy(|| format!("Failed to find FETCH_HEAD: {url}. This is most likely because the reference does not exist."))
```
Comment (shaunsingh) [inline src/smith/git.rs:180]: I'd probably change this error message, most people don't know what a reference is

**Review by shaunsingh — COMMENTED** [review]
added notes

File: src/smith/git.rs (hunk 3 above)
Comment (Suyashtnt) [reply src/smith/git.rs:180]: Something like "it is most likely because your git commit/tag/branch doesn't exist"?

File: src/manifest.rs (hunk 2 above)
Comment (Suyashtnt) [reply src/manifest.rs:260]: Will do

File: src/main.rs (hunk 1 above)
Comment (Suyashtnt) [reply src/main.rs:426]: Sure I'll do that

File: src/smith/git.rs (hunk 3 above)
Comment (shaunsingh) [reply src/smith/git.rs:180]: I think "Check if the specified commit, tag, or branch exists" sounds good

File: flake.nix [hunk 4]
```diff
@@ -70,9 +69,9 @@
       # artifacts from above.
       alpacka = craneLib.buildPackage {
         inherit cargoArtifacts src buildInputs;
+        doCheck = false;
```
Comment (shaunsingh) [inline flake.nix:72]: Not sure why we want to disable tests?

File: flake.nix [hunk 5]
```diff
@@ -105,11 +104,12 @@
         # Run tests with cargo-nextest
         # Consider setting `doCheck = false` on `my-crate` if you do not want
         # the tests to run twice
-        alpacka-nextest = craneLib.cargoNextest {
-          inherit cargoArtifacts src buildInputs;
-          partitions = 1;
-          partitionType = "count";
-        };
+        # Since the tests require internet access, nix can't run them
```
Comment (shaunsingh) [inline flake.nix:107]: Same thing here, I believe we can disable the sandbox altogether. We're ensuring our flake is pure anyways

**Review by shaunsingh — COMMENTED** [review]
Rust code looks good, just a few comments about the nix flake

File: flake.nix (hunk 4 above)
Comment (Suyashtnt) [reply flake.nix:72]: Don't rerun tests since we already have nexttest setup

File: flake.nix (hunk 5 above)
Comment (Suyashtnt) [reply flake.nix:107]: How can we disable the sandbox?

Comment (shaunsingh) [reply flake.nix:107]: The easiest way would be to run our tests with the `--no-sandbox` flag, though we could make it for the whole flake (`nix.useSandbox = false;`) if we wanted to.

Comment (Suyashtnt) [reply flake.nix:107]: > The easiest way would be to run our tests with the `--no-sandbox` flag, though we could make it for the whole flake (`nix.useSandbox = false;`) if we wanted to.

I think running the tests with `--no-sandbox` works best. Will do this later

**Pull request merged**